x-pack/legacy/plugins/maps/public/layers/vector_layer.js

```diff
@@ -484,12 +490,84 @@ export class VectorLayer extends AbstractLayer {
     }
   }
 
+  async _syncSourceFormatters(syncContext) {
+    if (this._style.constructor.type !== LAYER_STYLE_TYPE.VECTOR) {
+      return;
+    }
+
+    return this._syncFormatters({
+      source: this._source,
+      dataRequestId: SOURCE_FORMATTERS_ID_ORIGIN,
+      dynamicStyleProps: this._style.getDynamicPropertiesArray().filter(dynamicStyleProp => {
+        return dynamicStyleProp.getFieldOrigin() === FIELD_ORIGIN.SOURCE;
+      }),
+      ...syncContext,
+    });
+  }
+
+  async _syncJoinFormatters(syncContext, join) {
+    const joinSource = join.getRightJoinSource();
+    return this._syncFormatters({
+      source: joinSource,
+      dataRequestId: join.getSourceFormattersDataRequestId(),
+      dynamicStyleProps: this._style.getDynamicPropertiesArray().filter(dynamicStyleProp => {
+        const matchingField = joinSource.getMetricFieldForName(
+          dynamicStyleProp.getField().getName()
+        );
+        return dynamicStyleProp.getFieldOrigin() === FIELD_ORIGIN.JOIN && !!matchingField;
+      }),
+      ...syncContext,
+    });
+  }
+
+  async _syncFormatters({
+    source,
+    dataRequestId,
+    dynamicStyleProps,
```

Comment:
I would _not_ pass in dynamicStyleProps, since styles are not relevant really. I'd change this method-signature to accept a `fields`-array with `Field`-instances instead.

See more below